File: CPP/ByRound/2155/2155_D_Batteries.cpp
// Problem: CF 2155 D - Batteries
// https://codeforces.com/contest/2155/problem/D

/*
Problem: D. Batteries
Approach:
- This is an interactive problem where we need to find a working pair of batteries among n batteries.
- Some batteries work and some don't, but we know at least 2 work (a >= 2).
- We can test any two batteries in a flashlight which turns on if both work.
- Goal: Find two working batteries using at most floor(n^2 / a) queries.

Algorithm:
- The strategy uses a nested loop approach where we try pairs of batteries with increasing gap,
  starting from gap=1. For each fixed gap, we iterate through all batteries and test them against
  their "next" battery (with wraparound). When a pair works (returns 1), the solution stops.

Time Complexity: O(n^2) which is acceptable since n <= 40.
Space Complexity: O(1) - only using constant extra space.

*/

#include <stdio.h>
#include <iostream>

using namespace std;
#define int long long
#define deci long double
#define yes cout << "YES" << endl
#define no cout << "NO" << endl
#define inparr(n)                                                              \
  for (int i = 0; i < n; i++)                                                  \
    cin >> arr[i];
#define fr(s, n) for (int i = s; i < n; i++)
#define rfr(s, n) for (int i = s; i >= n; i--)
#define frj(s, n) for (int j = s; j < n; j++)
#define rfrj(s, n) for (int j = s; j >= n; j--)
#define vi vector<int>
#define pb push_back
#define vvi vector<vector<int>>
#define vpi vector<pair<int, int>>
#define mii map<int, int>
#define pii pair<int, int>
#define txtio                                                                  \
  freopen("substantial_losses_input.txt", "r", stdin);                         \
  freopen("output.txt", "w", stdout);
#define all(x) x.begin(), x.end()

void solve(int test) {
  int n;
  cin >> n;
  int gap = 1; // Start with gap of 1
  while (1) {
    for (int i = 1; i <= n; i++) { // Iterate over all batteries
      int nxt = i + gap; // Compute next battery index
      if (nxt > n)
        nxt %= n; // Wrap around if needed
      cout << i << ' ' << nxt << endl; // Query the flashlight with pair (i, nxt)
      cout.flush(); // Ensure output is flushed to prevent idleness limit error
      int x;
      cin >> x; // Read response from jury
      if (x) // If the flashlight turned on (both batteries work)
        return; // Solution found, exit
    }
    gap++; // Increase gap for next iteration
  }
}

signed main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int t = 1;
  cin >> t;
  for (int i = 1; i <= t; i++) {
    solve(i);
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Java/ByTier/D/2155_D_Batteries.java
// Problem: CF 2155 D - Batteries
// https://codeforces.com/contest/2155/problem/D

/*
 * Problem: D. Batteries
 * Purpose: Find a working pair of batteries among n batteries using at most floor(n^2 / a) queries,
 *          where a is the number of working batteries (unknown). The flashlight only turns on if both
 *          selected batteries work.
 *
 * Algorithm:
 * - Uses a cyclic search strategy to efficiently narrow down working pairs.
 * - Iterates through all possible distances from 1 to n/2.
 * - For each distance, loops over all starting positions and checks the pair (start, start + distance).
 * - If the result of a query is 1 or -1 (indicating a valid working battery pair), stop and return.
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 */
import java.util.Scanner;
public class D_Batteries {
    public static void main(String[] args) {
        Scanner alpha = new Scanner(System.in);
        int tests = alpha.nextInt(); // Read number of test cases
        while (tests-- > 0) {
            int n = alpha.nextInt(); // Read number of batteries
            executeCyclicSearch(n, alpha); // Perform search for working pair
        }
        alpha.close(); // Close scanner to free resources
    }
    
    static void executeCyclicSearch(int totalAlpha, Scanner scannerAlpha) {
        int halfAlpha = totalAlpha / 2; // Calculate maximum distance to test
        int distanceAlpha = 1; // Initialize current distance
        while (distanceAlpha <= halfAlpha) { // Iterate through all distances
            for (int startAlpha = 1; startAlpha <= totalAlpha; startAlpha++) { // Loop over start positions
                int endAlpha = startAlpha + distanceAlpha; // Compute end position
                endAlpha = (endAlpha > totalAlpha) ? endAlpha - totalAlpha : endAlpha; // Wrap around if necessary
                if (performQuery(startAlpha, endAlpha, scannerAlpha)) return; // Query and check result
            }
            distanceAlpha++; // Increase distance for next iteration
        }
    }
    
    static boolean performQuery(int firstAlpha, int secondAlpha, Scanner scannerAlpha) {
        System.out.println(firstAlpha + " " + secondAlpha); // Output query
        System.out.flush(); // Flush output to ensure immediate sending
        int responseAlpha = scannerAlpha.nextInt(); // Read response from jury
        return responseAlpha == 1 || responseAlpha == -1; // Return true if found a working pair
    }
}


// https://github.com/VaHiX/CodeForces/